### ISSUE TITLE
Repository.save executes SELECT before UPDATE

### ISSUE THREAD

**Issue by the-owl** [typeorm/typeorm#3772]
**Issue type:**

[x] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[x] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

When an existing entity is saved into repository, it runs `SELECT` first and runs `UPDATE` only after that. This makes `.save()` slower than `.update()`. This issue (or not an issue?) is reproducible with PostgreSQL. I don't know if it's reproducible with other supported databases.

The following code can be used to reproduce this behavior:
```ts
import 'reflect-metadata';
import { createConnection, Entity, PrimaryGeneratedColumn, Column } from 'typeorm';


@Entity()
export class TestEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  someData: number;

  constructor(data: number) {
    this.someData = data;
  }
}

async function main () {
  const connection = await createConnection({
    type: 'postgres',
    host: 'localhost',
    port: 5432,
    username: 'user',
    password: '123',
    database: 'db',
    synchronize: true,
    entities: [TestEntity],
    logging: true
  });
  const repo = connection.getRepository(TestEntity);
  await repo.insert(new TestEntity(1));

  const entity = (await repo.findOne({}))!;

  console.log('-------------')

  entity.someData = 2;
  await repo.save(entity);
}

main();
```

These logs show that `SELECT` is triggered by `.save()` before `UPDATE`.
```
...
query: INSERT INTO "test_entity"("someData") VALUES ($1) RETURNING "id" -- PARAMETERS: [1]
query: SELECT "TestEntity"."id" AS "TestEntity_id", "TestEntity"."someData" AS "TestEntity_someData" FROM "test_entity" "TestEntity"
-------------
query: SELECT "TestEntity"."id" AS "TestEntity_id", "TestEntity"."someData" AS "TestEntity_someData" FROM "test_entity" "TestEntity" WHERE ("TestEntity"."id" IN (1))
query: START TRANSACTION
query: UPDATE "test_entity" SET "someData" = $1 WHERE "id" IN (1) -- PARAMETERS: [2]
query: COMMIT
```

Is this behavior intentional? If this is not a bug, what is the reason for running `SELECT` before `UPDATE`? Also, if this is not a bug, I think that this behavior should be noted in documentation.

**Comment by johannesschobel** [comment]
i guess, this is, because `save()` does not expect an `id` param, like, for example, the `update()` method requires. Therefore, `TypeORM` first needs to resolve the id..
The SQL SELECT statement tries resolving the passed object and then use the `id` param to fire the second UPDATE statement.

**Comment by the-owl** [comment]
> i guess, this is, because `save()` does not expect an `id` param, like, for example, the `update()` method requires. Therefore, `TypeORM` first needs to resolve the id..

No, TypeORM already knows `id` of the entity. It executes `SELECT` with condition `id = 1`:
```
...
query: SELECT "TestEntity"."id" AS "TestEntity_id", "TestEntity"."someData" AS "TestEntity_someData" FROM "test_entity" "TestEntity" WHERE ("TestEntity"."id" IN (1))
...
```

**Comment by omnitired** [comment]
that's because Repository.save always checks if the entity exists or not, if it does exist it will update it and if not, if will insert it. that's just how it works

**Comment by the-owl** [comment]
> that's because Repository.save always checks if the entity exists or not, if it does exist it will update it and if not, if will insert it. that's just how it works

It could just execute UPDATE without any existence checks. If entity is already removed, UPDATE will not fail, it will just do nothing, so I see no reason for doing SELECT first.

**Comment by johannesschobel** [comment]
The current situation works on either DB.. your suggestion may be specific to certain implementations..

**Comment by vlapo** [comment]
This behaviour is by design. Sure your example is just a simple update but `save` method do more: update if entity exist in database, insert if not, run listeners, support partial updates, handling relations and more. To perform more advanced functionalities we need to get actual data from DB. Feel free to check implementation https://github.com/typeorm/typeorm/blob/0aa79173e667e193cee653492176a43eb06abcd5/src/persistence/EntityPersistExecutor.ts

**Comment by neuro-sys** [comment]
I see that this is by design, but please consider this as an improvement or question:

1) In a scenario when an Entity is fetched fully by the application, then modified partially, and before saving, EntityPersistExecutor fetches it fully once _again_ (as stated, by design). Re-fetching makes sense when the original entity was not fully selected. But if it's already a fully "hydrated" entity, there is no reason to fetch the same data again (except when the initial select does not occur inside an appropriate transaction type). At least this could be provided as an option where the default behaviour would be as it is right now, which otherwise re-use the existing Entity as a subject for the SubjectExecutor to do its job.

2) Is it possible to access the entities loaded by SubjectDatabaseEntityLoader? This could be useful in scenarios when the entity is partially loaded by the user.

**Comment by OLDIN** [comment]
any updates?
I have same problem.

**Comment by anatolyukropov** [comment]
Yes, got this probmel too. I need bulk insert 500 rows. And this get very very very... long time.

**Comment by dsbert** [comment]
You can use `.insert` for quick, direct inserts operations. It  won't check for an existing entity.

https://typeorm.io/#/repository-api

**Comment by dougmaitelli** [comment]
I completely agree with @vlapo, this is not an issue, in many cases you want to "insert or update" a record and save does exactly that. But I also understand the reason behind this discussion, in many cases people are just persisting a pre-fetched entity as mentioned by @neuro-sys.

Probably the best approach for typeorm is to implement some sort of internal handling to understand if the entity was already previously fetched or even another sort of overload for .save().

Also, I would like to point out, that, given the fact `.save()` is designed to be a "insert or update" sort of method it is very strange that you can't just use `.save({username: "test", name: "John", lastName: "Doe"})` to insert a new record because save will complain about missing properties that would only be available to pre-fetched entities.

So basically `.save()` does a select before update because it does "insert or update" but you can't use it for insert because it requires properties that are not present in a non-fetched entity.

**Comment by Shivam7478** [comment]
@the-owl  @neuro-sys @dougmaitelli @dsbert @johannesschobel @vlapo @
Please use reload:false & transaction:false parameter with the query 

**for example:-**

    ```
  await this.save(addUserDto, {
        reload: false,
        transaction: false,
      });
```

[Please check here other options](https://typeorm.delightful.studio/interfaces/_repository_saveoptions_.saveoptions.html)

**Comment by dougmaitelli** [comment]
Thanks for the info @Shivam7478, I think that address the original request from @the-owl.
But there is still the problem I pointed out. For some reason if I try to do a `.save({username: "test", name: "John", lastName: "Doe"})` typescript complains that there are many missing fields in the DTO (which are actually internal metadata fields from typeorm.

**Comment by Mati365** [comment]
> @the-owl @neuro-sys @dougmaitelli @dsbert @johannesschobel @vlapo @
> Please use reload:false & transaction:false options with the query
> 
> **for example:-**
> 
> ```
>   await this.save(addUserDto, {
>         reload: false,
>         transaction: false,
>       });
> ```
> 
> [Please check here other options](https://typeorm.delightful.studio/interfaces/_repository_saveoptions_.saveoptions.html)

it doesn't work

**Comment by r1cardosouza** [comment]
The SELECT before UPDATE only vanish when I use update instead of save, `reload: false` not worked for me also.
So, I did as follow:

await MyModel.update (
    {Id: dbModel.Id},
    dbModel
  );

**Comment by ganshiqingyuan** [comment]
> The SELECT before UPDATE only vanish when I use update instead of save, `reload: false` not worked for me also.
> So, I did as follow:
> 
> await MyModel.update (
> {Id: dbModel.Id},
> dbModel
> );

this maked all the field be updated,,in fact you should only update what you want to update maybe like this
```
const changedEntity: Partial<MessageStatus> = {
                isShow: IsShow.show,
                lastMessageContent: messageContent,
                lastMessageTime: new Date(),
                messageNums: ownMessageStatus.messageNums + 1
            }
            Object.assign(ownMessageStatus, changedEntity)
            this.ownMessageStatusRepostory.update(ownMessageStatus.id, changedEntity).then(res=>{
               return ownMessageStatus
            })
```

**Comment by xtance** [comment]
And also because of this @ Afterload gets called after updating a single property in an entity

**Comment by Yuvraj102** [comment]
I'm having somewhat similar issue updating record using typeorm's repository 
https://stackoverflow.com/questions/74068602/typeorm-update-method-doesnt-work-when-using-repository

**Comment by dberardo-com** [comment]
is it possible ot use insert() but returinging the inserted entities ? i currently use save() but that requieres 2 queries. i need entities as return, not raw stuff

**Comment by alumni** [comment]
@dberardo-com it depends on the database that you use. Some databases support "insert ... returning". See this example from the tests:

https://github.com/typeorm/typeorm/blob/5003aaa7c5143a4a718f1d9c3de227cc12e19dff/test/functional/query-builder/returning/query-builder-returning.ts#L31-L40

https://github.com/typeorm/typeorm/blob/5003aaa7c5143a4a718f1d9c3de227cc12e19dff/test/functional/query-builder/returning/query-builder-returning.ts#L57-L60

`returning.raw` will contain the raw results that the database driver is returning.

Besides Postgres (and alike), SQL Server and Spanner, I think it's also supported in MySQL/MariaDB and Oracle. I remember seeing a PR adding support for SQLite-like drivers too, so that might come in the future if the PR gets cleaned up.

You can maybe try it out.

**Comment by dberardo-com** [comment]
I am using postgres, but the return statement is not what I mean, the issue here is that I am inserting complex objects and typeorm makes many inserts under the hood, so I would like to have the complete entity with all relations etc.... I mean exactly like .save() 

This is something I cannot achieve with insert on the query builder.

Does it make sense or shall I make it more clear?
